src/uri.rs

```diff
@@ -176,6 +176,77 @@ impl fmt::Display for RsyncModule {
 }
 
 
+//------------ Http ----------------------------------------------------------
+
+/// A simple HTTP(s) URI
+///
+/// This supports only what we need for the references in RPKI objects and
+/// publication / provisioning messages. In particular, this does not support
+/// the query and fragment components of URIs.
+#[derive(Debug, PartialEq)]
+pub struct Http {
+    scheme: Scheme,
+    host:   Bytes,
+    path:   Bytes
+}
+
+impl Http {
+
+    pub fn from_slice(slice: &[u8]) -> Result<Self, Error> {
+        Self::from_bytes(slice.into())
+    }
+
+    pub fn from_bytes(mut bytes: Bytes) -> Result<Self, Error> {
+        if !is_uri_ascii(&bytes) {
+            return Err(Error::NotAscii)
+        }
+
+        let scheme = scheme(&bytes)?;
+
+        bytes.advance(
+            match scheme {
+                Scheme::Http => { 7 }
+                Scheme::Https => { 8 }
+        });
+
+        let host_length = {
+            let mut parts = bytes.splitn(3, |ch| *ch == b'/');
+            match parts.next() {
+                Some(host) => { host.len() }
+                None => return Err(Error::BadUri)
+            }
+        };
+
+        let host = bytes.split_to(host_length);
+        let path = bytes;
+
+        if path.len() == 0 {
+            return Err(Error::BadUri)
+        }
+
+        Ok(Http{scheme, host, path})
+    }
+
+}
+
+fn scheme(bytes: &Bytes) -> Result<Scheme, Error> {
```

Comment:
This should probably be an associated function of `uri::Http` (i.e., live inside the `impl Http`), since it is very specific to HTTP URIs. Maybe also call it `take_scheme` or something it let it advance the `bytes` directly?

Reply:
I made this an associated function of `uri::Scheme` instead and let both Rysnc and Http use this to get the Scheme out. I changed it to `take_scheme`, good suggestion.